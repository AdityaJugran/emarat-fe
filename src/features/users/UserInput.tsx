import { Dispatch, SetStateAction } from 'react';
import { Modal, Spin } from 'antd';
import createFormProps from 'features/users/formProps';
import GenericForm from 'features/shared/components/form/GenericForm';
import { UserType } from 'features/users/Types';
import { useApiCall } from 'config/hooks';
import { addPrefetchOptions } from 'features/shared/components/form/genericFormHelper';
import SpinContainer from 'features/shared/components/styledComponents/SpinContainer';

export type PropsType = {
	isVisible: boolean;
	setIsVisible: Dispatch<SetStateAction<boolean>>;
	edit?: UserType | null;
	submitCallback?: () => void;
};

export default function UserInput({
	isVisible,
	setIsVisible,
	edit,
	submitCallback,
}: PropsType) {
	const { data: amenitiesData, loading } = useApiCall({
		apiUrl: 'amenities',
		initDataValue: [],
		appendToUrl: '?basic=no',
	});
	const formSubmitCallback = (data: UserType) => {
		setIsVisible(false);
		submitCallback && submitCallback();
	};

	const formData = edit
		? {
				...createFormProps,
				meta: { ...createFormProps.meta, apiUrl: 'putUser' },
		  }
		: createFormProps;

	const updateValues = edit
		? { ...edit, isAdmin: edit.isAdmin ? 'true' : 'false' }
		: undefined;

	const formDataWithAmenities = addPrefetchOptions({
		field: 'amenities',
		form: formData,
		options: amenitiesData,
		labelKey: 'name',
		valueKey: '_id',
	});

	return (
		<Modal
			visible={isVisible}
			okText="Create"
			footer={null}
			onCancel={() => setIsVisible(false)}
			centered
		>
			<h2 className="text-2xl mb-4">{edit ? 'Edit' : 'Add'} User</h2>
			{isVisible &&
				(loading ? (
					<SpinContainer>
						<Spin />
					</SpinContainer>
				) : (
					<GenericForm
						appendToUrl={edit?._id}
						updateValues={updateValues || undefined}
						formData={formDataWithAmenities}
						layout="vertical"
						submitCallback={formSubmitCallback}
					/>
				))}
		</Modal>
	);
}
