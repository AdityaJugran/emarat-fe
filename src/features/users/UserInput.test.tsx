import renderer from 'react-test-renderer';
import UserInput from 'features/users/UserInput';

it('Login Form', () => {
	const tree = renderer
		.create(<UserInput isVisible={false} setIsVisible={jest.fn()} />)
		.toJSON();
	expect(tree).toMatchSnapshot();
});
