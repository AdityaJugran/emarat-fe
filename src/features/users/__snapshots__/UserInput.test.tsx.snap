// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Form 1`] = `null`;
